### PR TITLE
Rewrite rebasing in `jj split` using `transform_descendants()`

### PR DESCRIPTION
**Pull request by emesterhazy**
This is following on the rewrite for `parallelize`.

- https://github.com/martinvonz/jj/pull/3521

Since rebase_descendants from rebase.rs is no longer used outside of that file, it can be made private again.

### CODE REVIEW

File: cli/src/commands/split.rs [hunk 1]
```diff
@@ -179,16 +174,33 @@ the operation will be aborted.
...
+            for parent in old_parents {
+                if parent != commit.id() {
+                    new_parents.push(parent.clone());
+                    continue;
+                }
+                if args.siblings {
+                    new_parents.push(first_commit.id().clone());
+                }
+                new_parents.push(second_commit.id().clone());
+                num_rebased += 1;
```
Comment (martinvonz) [inline cli/src/commands/split.rs:194]: Shouldn't this be updated outside the `for parent in old_parents` loop?

File: cli/src/commands/split.rs [hunk 2]
```diff
@@ -179,16 +174,33 @@ the operation will be aborted.
...
-    };
+    let mut num_rebased = 0;
+    tx.mut_repo().transform_descendants(
+        command.settings(),
+        vec![commit.id().clone()],
+        |mut rewriter| {
+            let old_parents = rewriter.old_commit().parent_ids();
+            // Reserve space up front since we know we're adding at most one more parent.
+            let mut new_parents = Vec::with_capacity(old_parents.len() + 1);
+            for parent in old_parents {
```
Comment (martinvonz) [inline cli/src/commands/split.rs:185]: We should probably add some more helper to `CommitRewriter` to make it easier to do this kind of transformation. We can do that later.

File: cli/src/commands/split.rs [hunk 3]
```diff
@@ -179,16 +174,33 @@ the operation will be aborted.
...
+                    continue;
+                }
+                if args.siblings {
+                    new_parents.push(first_commit.id().clone());
+                }
+                new_parents.push(second_commit.id().clone());
+                num_rebased += 1;
+            }
+            rewriter.set_new_rewritten_parents(new_parents);
+            if rewriter.parents_changed() {
```
Comment (martinvonz) [inline cli/src/commands/split.rs:197]: This is always true, isn't it? So it's probably clearer to drop it.

**Review by martinvonz — APPROVED** [review]
Thanks! I'm happy to see `rebase_descendants()` made private again. That will make it easier for @bnjmnt4n to clean that up later.

I'm approving this because it seems clearly not specific to google, but please give non-googlers a chance to review.

File: cli/src/commands/split.rs (hunk 3 above)
Comment (emesterhazy) [reply cli/src/commands/split.rs:197]: It's only true if we actually changed the parents, right? If we split a commit that has a lot of descendants won't this visit all of them, most of which won't be rebased?

File: cli/src/commands/split.rs (hunk 2 above)
Comment (emesterhazy) [reply cli/src/commands/split.rs:185]: Ack.

File: cli/src/commands/split.rs (hunk 3 above)
Comment (martinvonz) [reply cli/src/commands/split.rs:197]: It would visit all of them, but since the root commit (the split one) was rewritten, all descendants would have to be rewritten too, right?

File: cli/src/commands/split.rs (hunk 1 above)
Comment (emesterhazy) [reply cli/src/commands/split.rs]: Actually yeah, I think so. I misunderstood how the count worked previously. I pushed another commit with testing changes that would have caught this.

File: cli/src/commands/split.rs (hunk 2 above)
Comment (martinvonz) [reply cli/src/commands/split.rs]: Actually, the existing `replace_parent()` should already work well here. I think this should work:

```rust
if args.siblings {
    rewriter.replace_parent(second_commit.id(), &[first_commit.id().clone(), second_commit.id().clone()]);
}
```

File: cli/src/commands/split.rs (hunk 3 above)
Comment (emesterhazy) [reply cli/src/commands/split.rs:197]: Doesn't that apply to parallelize as well?

Comment (martinvonz) [reply cli/src/commands/split.rs:197]: > Doesn't that apply to parallelize as well?

No, because `parallelize` leaves the first commit in the set unchanged (after I added this check), which means that descendants can also be left unchanged if they depend only on that one commit (e.g. when you do `jj parallelize 'A|C' in a chain of commits A..D`). But `jj split` *always* rewrites the old commit. We could make it not do that if no paths went into the second commit (e.g. `jj split -r <commit> 'all()'`) but it's much less useful for `jj split` than it is for `jj parallelize`.

Comment (emesterhazy) [reply cli/src/commands/split.rs:197]: Ok I dug through the code and figured out my misunderstanding. I didn't realize that the new parents of the commit are being set by `transform_descendants` taking into account previously rewritten commits even if I don't do anything in the callback.

https://github.com/martinvonz/jj/blob/77eaf67f96c6ea438e2dd1a5636f34b6a321e6be/lib/src/repo.rs#L1175-L1178

https://github.com/martinvonz/jj/blob/77eaf67f96c6ea438e2dd1a5636f34b6a321e6be/lib/src/repo.rs#L924-L929

And this takes into account any commits that have already been rewritten, so every descendant (in the case of `split`) has different parents and the if condition is always true.

Comment (emesterhazy) [reply cli/src/commands/split.rs:197]: The only thing I think I'm still confused about is how I wasn't messing up the `new_parents` Vec in the rewriter. If the parent IDs in that Vec took into account previous rewrites, then it seems like the parents should have been different than the ones I took pulled from the old commit.

In fact, that appears to be true because when I rewrote this to use `replace_parent` it didn't work anymore since the original commit ID isn't present in the `new_parents` vector. It's already been replaced by the ID of the new `second_commit` from when we called `set_rewritten_commit` before running the rebaser.

Can you help me understand this part? I assume after updating the `new_parents` Vec with old commit IDs that were rewritten eventually the old IDs were mapped to the new IDs, but I'm not sure.

File: cli/tests/test_split_command.rs [hunk 4]
```diff
@@ -346,8 +346,8 @@ fn test_split_siblings_no_descendants() {
     let (stdout, stderr) = test_env.jj_cmd_ok(&workspace_path, &["split", "--siblings", "file1"]);
     insta::assert_snapshot!(stdout, @"");
     insta::assert_snapshot!(stderr, @r###"
-    First part: qpvuntsm?? 8d2b7558 TESTED=TODO
```
Comment (emesterhazy) [inline cli/tests/test_split_command.rs]: I'm not sure why this was marked with a `??` before or why it's gone now. I guess the change became divergent somehow due to a bug in the old rebasing logic?

File: cli/src/commands/split.rs (hunk 2 above)
Comment (emesterhazy) [reply cli/src/commands/split.rs]: Yep, this works. We don't need to do anything for the non-siblings case since we already marked the original commit as rewritten.

File: cli/tests/test_split_command.rs (hunk 4 above)
Comment (emesterhazy) [reply cli/tests/test_split_command.rs]: Looks like it was only divergent at the time when we printed the summary and it got fixed later during `tx.finish()` which calls `rebase_descendants()`.

The bug is that if the commit has no children then we don't call `rebase_descendants()` when `--siblings` is used.

https://github.com/martinvonz/jj/blob/77eaf67f96c6ea438e2dd1a5636f34b6a321e6be/cli/src/commands/split.rs#L223-L235

That's fixed after this PR.

**Review by yuja — APPROVED** [review]

File: cli/src/commands/split.rs (hunk 3 above)
Comment (martinvonz) [reply cli/src/commands/split.rs]: > Can you help me understand this part? I assume after updating the `new_parents` Vec with old commit IDs that were rewritten eventually the old IDs were mapped to the new IDs, but I'm not sure.

I don't think that would happen. Are you asking about why some previous version of the code worked when you expected it not to? Can you point to which commit that was?

Comment (emesterhazy) [reply cli/src/commands/split.rs]: Sorry! I think what I wrote above was pretty hard to follow. In my first draft (https://github.com/martinvonz/jj/commit/6da4e9118f8f02a5a73b511368dddeeff1024bbf) I was taking the parent IDs from the old commit, putting them into a vector, and calling `set_new_rewritten_parents` with it.

- https://github.com/martinvonz/jj/blob/77eaf67f96c6ea438e2dd1a5636f34b6a321e6be/lib/src/rewrite.rs#L148-L152

In my head I confused this with `set_new_parents`, which doesn't take rewrites into account. I think I understand the API a little better now. Thanks for being patient :)

**Comment by emesterhazy** [comment]
Thanks for taking a look Yuya. I'll merge this now.
